### PR TITLE
Analyzer: filtering by virtual columns for StorageS3

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 3256a000c2a0b3e9af3a0ac8c9a3dd60e6260828 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56668/3256a000c2a0b3e9af3a0ac8c9a3dd60e6260828/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6849490086">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56668/3256a000c2a0b3e9af3a0ac8c9a3dd60e6260828/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Storages/StorageS3.cpp [hunk 1]
```diff
@@ -125,6 +127,138 @@ namespace ErrorCodes
...
+
+    Block header = getBlockWithVirtuals(virtual_columns, "", {});
+
+    const auto & table_expression_data = query_info.planner_context->getTableExpressionDataOrThrow(query_info.table_expression);
+    for (const auto & [column_name, column_identifier] : table_expression_data.getColumnNameToIdentifier())
+        column_rename.emplace(column_name, column_identifier);
+
+    header = renameColumnsInBlock(header, column_rename);
+    auto filter_dag = buildActionsDAGFromExpressionNode(
+        query_node->getWhere(), header.getColumnsWithTypeAndName(), query_info.planner_context);
```
Comment (novikd) [inline src/Storages/StorageS3.cpp:216]: This looks very suspicious. You should not extract filter from the query tree. Filter must be stored in the `SelectQueryInfo::filter_actions_dag`.
`PartitionPruner` uses `SelectQueryInfo::filter_actions_dag`.

**Review by novikd — COMMENTED** [review]
The code around filter DAG creation seems to be incorrect.

**Review by novikd — APPROVED** [review]

**Comment by vdimir** [comment]
> Integration tests (tsan) [4/6] — fail: 1, passed: 422 [Details](https://s3.amazonaws.com/clickhouse-test-reports/56668/31a6c7c1c49da5f9ea42f97bee7e617cebabaebe/integration_tests__tsan__[4_6].html)

`test_storage_rabbitmq`

> Stress test (tsan) — Sanitizer assert (in stderr.log) [Details](https://s3.amazonaws.com/clickhouse-test-reports/56668/31a6c7c1c49da5f9ea42f97bee7e617cebabaebe/stress_test__tsan_.html)

https://github.com/ClickHouse/ClickHouse/issues/57009

**Comment by Avogar** [comment]
I don't understand these changes. New filtering with `filterKeysForPartitionPruning` is used only before creating `KeysIterator`, but this iterator in most cases contains only 1 key, because for urls with globs we use `DisclosedGlobIterator`, and you didn't change the filtering in this iterator, so filtering still doesn't work when we have StorarteS3 with url that contains globs (and it's the main use case of filtering).
Also interesting why with analyzer filtering works with s3 table function withotu these changes and doesn't work with table created with `S3` engine

File: src/Storages/StorageS3.cpp [hunk 2]
```diff
@@ -1091,15 +1193,34 @@ Pipe StorageS3::read(
     size_t max_block_size,
     size_t num_streams)
 {
-    auto query_configuration = updateConfigurationAndGetCopy(local_context);
+    auto read_from_format_info = prepareReadingFromFormat(column_names, storage_snapshot, supportsSubsetOfColumns(local_context), virtual_columns);
```
Comment (Avogar) [inline src/Storages/StorageS3.cpp:1196]: This `read_from_format_info` is used only for getting `source_header`, and then we call `prepareReadingFromFormat` again in `ReadFromStorageS3Step::initializePipeline`. Why don't just pass `read_from_format_info` as an argument in `ReadFromStorageS3Step`?

**Comment by Avogar** [comment]
```sql
:) create table test (number UInt64) engine=S3('http://localhost:11111/test/data*.jsonl');
:) select count(), _file from test where _file='data1.jsonl' group by _file settings allow_experimental_analyzer = 1;

┌─count()─┬─_file───────┐
│      10 │ data1.jsonl │
└─────────┴─────────────┘

[avogar-dev] 2023.12.29 13:49:56 [ 0 ] EngineFileLikeReadFiles: 2 (increment)
```

So we still read 2 files instead of 1. But when we do the same using table function, it works:
```

:) select count(), _file from s3('http://localhost:11111/test/data*.jsonl') where _file = 'data1.jsonl' group by _file settings allow_experimental_analyzer=1

┌─count()─┬─_file───────┐
│      10 │ data1.jsonl │
└─────────┴─────────────┘
[avogar-dev] 2023.12.29 13:51:49 [ 0 ] EngineFileLikeReadFiles: 1 (increment)
```

What is the difference between reading from storage and table function in analyzer?

**Comment by Avogar** [comment]
And should we do smth similar with other engines like `File, URL, HDFS, AzureBlob` that supports filtering

**Comment by Avogar** [comment]
And also new functions `getBlockWithVirtuals/filterKeysForPartitionPruning` duplicate some logic from `VirtualColumnUtils::filterByPathOrFile/getFilterByPathAndFileIndexes`. As I understand the new approach with filter using `ActionsDAG` should work with both current infrastructure and new analyzer, why not to rewrite all filtering in `S3/URL/HDFS/File/Azure` to new approach? We will need to just rewrite common functions in `VirtualColumnUtils` for using and use them in that engines (most filtering functions from `VirtualColumnUtils` are used only in these engines).
Right now this new code in single enging for single use case with `KeysIterator` looks really strange

**Pull request merged**